### ISSUE TITLE
autocovariance in convergence.R gives incorrect output

### ISSUE THREAD

**Issue by zacmon** [stan-dev/posterior#126]
Hi,

Thanks so much for writing this package. I'm currently adapting convergence.R to python and pandas to meet my needs and uses. I was originally using [monitor.R](https://github.com/stan-dev/rstan/blob/develop/rstan/rstan/R/monitor.R) in RStan, which has a GPL3 license. I noticed that convergence.R has been maintained more recently, so I've taken a look at that. Of course, the two are similar, but posterior doesn't have a license. Will it also be GPL 3? (Perhaps this should be a separate issue?)

I am running checks on my adaptation to ensure that my code output matches that of convergence.R. When I compare output from my script and convergence.R by analyzing chains from [here](https://avehtari.github.io/rhat_ess/rhat_ess.html) or analyzing chains of any set of random numbers, the two match. However, I sought to test some even simpler cases. E.g. I wanted to pretend a range of 0 to 100 was a chain and compute the effective sample size. The autocovariance function in covergence.R gives unexpected results, compared to what I'm using in [statsmodels in Python](https://www.statsmodels.org/stable/generated/statsmodels.tsa.stattools.acovf.html) and what I compute without the use of FFTs.

<img width="583" alt="Screen Shot 2021-05-04 at 10 33 12 PM" src="https://user-images.githubusercontent.com/7682998/117100520-c62bfa00-ad28-11eb-85d0-b1640597729c.png">

whereas the autocovariance function using statsmodels gives

<img width="604" alt="Screen Shot 2021-05-04 at 10 34 36 PM" src="https://user-images.githubusercontent.com/7682998/117100580-ef4c8a80-ad28-11eb-869b-5f66b0ce0dbf.png">

Notably, the first entry in the autocovariance should be the variance (with N in the divisor), which comes out to 850 for a range from 0 to 100. It's not clear to me why the `autocovariance` function in convergence.R doesn't calculate this correctly.

I tested ranges of 0 to 1, 2, 3, 4, 5, 6, 7, 8, and 9, and both functions matched. It appears to be that only after a range from 0 to at least 10 does the R autocovariance function differ in output.

Any thoughts as to why this might be?

Thanks,
Zach

**Comment by jgabry** [comment]
Sorry for the slow reply! 

> Of course, the two are similar, but posterior doesn't have a license. Will it also be GPL 3? (Perhaps this should be a separate issue?)

Actually there is a license. It's BSD-3, which is similar to MIT and is less restrictive than GPL: 

https://github.com/stan-dev/posterior/blob/master/LICENSE.md

So I think you're free to do pretty much anything you want with it as long as you give credit. 

> Notably, the first entry in the autocovariance should be the variance (with N in the divisor), which comes out to 850 for a range from 0 to 100. It's not clear to me why the `autocovariance` function in convergence.R doesn't calculate this correctly.

I'm not sure. I think @avehtari and/or @paul-buerkner wrote that function so hopefully one of them can tell you if there's a mistake or if the behavior is intentional.

**Comment by paul-buerkner** [comment]
I think @avehtari wrote the initial function. I may have edited it at some point but I wouldn't claim to understannd all the details.

**Comment by avehtari** [comment]
Thanks for the report. 
The difference comes from our function using zero padded FFT and statsmodels not using zero padding. Use of zero padding was copied from Stan C++. Bob Carpenter wrote in blog post 2012 that the zero padding had a big effect on speed https://lingpipe-blog.com/2012/06/08/autocorrelation-fft-kiss-eigen/. Zero padding the size of input to be a power of 2, 3, or 5 is usually recommended for speed (due to how FFT works). I just tested that R's FFT has also a big speed difference for a long input with or without padding, so the padding should stay. Without padding R's fft() can be x100 slower for input lengths that are reasonable for Stan users. However it seems we should add mask correction https://etudes.tibonihoo.net/literate_musing/autocorrelations.html to remove the possible error from zero masking. 

There is no hurry to add this as this kind of clearly non-stationary sequences are very unlikely and even then, e.g., ESS estimates are close enough to 1 so the error will not make anyone to think their chains converged when they didn't. So the fix would be mostly for purity and to people spending time wondering why the output doesn't match the true value (even if strange input is used).

Here's the code I think is correct
```
autocovariance <- function(x) {
  N <- length(x)
  M <- fft_next_good_size(N)
  Mt2 <- 2 * M
  yc <- x - mean(x)
  yc <- c(yc, rep.int(0, Mt2 - N))
  transform <- fft(yc)
  ac <- fft(abs(transform)^2, inverse = TRUE)
  # mask correction
  (mask <- c(rep.int(1, N), rep.int(0, Mt2 - N)))
  mask_correction <- fft(abs(fft(mask))^2, inverse = TRUE)
  # use "biased" estimate as recommended by Geyer (1992)
  # without mask correction
  # ac <- Re(ac)[1:N] / (N^2 * 2)
  # with mask correction
  ac <- Re(ac)[1:N] / Re(mask_correction[1])
  ac
}
```

This can be tested against `acf(,type="covariance", lag.max=N)`, which seems to use FFT without zero padding (based on the speed behavior)

**Comment by jgabry** [comment]
Thanks for the explanation Aki!

**Comment by avehtari** [comment]
We can avoid the need to compute the mask effect (as it requires additional FFT and IFFT computations)
This works as it's enough that we scale the autocovariances so that the first element matches var(x)*(N-1)/N
```
autocovariance <- function(x) {
  N <- length(x)
  # zero padding makes fft() much faster when N>1000
  M <- fft_next_good_size(N)
  Mt2 <- 2 * M
  yc <- x - mean(x)
  yc <- c(yc, rep.int(0, Mt2 - N))
  # FFT based unnormalized autocovariances
  ac <- Re(fft(abs(fft(yc))^2, inverse = TRUE)[1:N])
  # use "biased" estimate as recommended by Geyer (1992)
  # direct scaling with var(x) avoids need to compute "mask effect"
  ac <- ac / ac[1] * var(x)*(N-1)/N
}
```

**Comment by paul-buerkner** [comment]
@avehtari is there anything we should add or change in posterior in response to this issue? Otherwise, I would close it now.

**Comment by rok-cesnovar** [comment]
I have a few suggestions with regards to licensing to avoid confusion in the future: 

- add a short note in PRs that the copyright holder is agreeing to license the submitted work. 
This is already a given as Github rules state that the copyright holder is agreeing to license the submitted work under the licenses of the repository, but I think its better to state it clearly to avoid any confusion later.

![image](https://user-images.githubusercontent.com/28476796/119271073-3b8c3b80-bc00-11eb-9fd1-7582e79a47cd.png)

- add a note in the Readme.md

- fix the license file to include the license name https://github.com/stan-dev/posterior/blob/master/LICENSE 

This would make the license explicit on the repository front page (example for https://github.com/stan-dev/math)

![image](https://user-images.githubusercontent.com/28476796/119271481-07b21580-bc02-11eb-82ae-ae9cc412026b.png)

For the first 2 I have the changes on a branch https://github.com/stan-dev/posterior/tree/pr_licenses

**Comment by jgabry** [comment]
I think that all sounds good except unfortunately this:

> * fix the license file to include the license name https://github.com/stan-dev/posterior/blob/master/LICENSE

I wish we could do that but I think the change we actually need to make is to reformat it to conform to CRAN policies (right now we're not even doing that). For a non-standard license like BSD (I don't know why CRAN considers BSD and MIT non-standard) they require a LICENSE file that has only the following lines: 

```
YEAR: ...
COPYRIGHT HOLDER: ...
```

and then the actual name of the license goes in DESCRIPTION. 

Unfortunately that means that GitHub will not automatically detect the license from the file (this is a well know annoyance in the R community) and it's why R packages like [ggplot2](https://github.com/tidyverse/ggplot2) that use the MIT license have two license files, LICENSE (which is just the two lines above for CRAN) and LICENSE.md (the actual license file). We also have this for CmdStanR in anticipation of CRAN.

**Comment by rok-cesnovar** [comment]
Ah ok, thanks for the explanation. Makes sense then. Retracting that suggestion then :)

**Comment by jgabry** [comment]
> Makes sense then.

I wish the CRAN policy also made sense ;)

**Comment by rok-cesnovar** [comment]
Yeah, your explanation makes sense. CRANs... less so. :)

**Comment by avehtari** [comment]
> @avehtari is there anything we should add or change in posterior in response to this issue? Otherwise, I would close it.

Unfortunately there are two topics in one issue. 1) I think the license issue is solved. 2) I think the autocovariance issue is not solved. 

I posted the autocorrelation code in hope someone else would have a look before progressing (you should not trust my conclusions or code). In addition the code should not differ from Stan math code https://github.com/stan-dev/math/blob/92075708b1d1796eb82e3b284cd11e544433518e/stan/math/prim/fun/autocorrelation.hpp and https://github.com/stan-dev/math/blob/92075708b1d1796eb82e3b284cd11e544433518e/stan/math/prim/fun/autocovariance.hpp, So the possible difference between these should be checked. It looks like Stan math is doing the correct thing that corresponds to the new code I posted, and I assume that at some point the R code was simplified but then skipped one step assuming it would be not needed while it was needed.  I assume previously the difference between different implementations was small as no-one tested these with such an extreme example as in this issue.

**Comment by paul-buerkner** [comment]
Thanks Aki. Let's move the licence discussion to a new issue then (if there is more to discuss) and keep this issue here for the autocovariance stuff.

**Comment by zacmon** [comment]
> Thanks for the report.
> The difference comes from our function using zero padded FFT and statsmodels not using zero padding. Use of zero padding was copied from Stan C++. Bob Carpenter wrote in blog post 2012 that the zero padding had a big effect on speed https://lingpipe-blog.com/2012/06/08/autocorrelation-fft-kiss-eigen/. Zero padding the size of input to be a power of 2, 3, or 5 is usually recommended for speed (due to how FFT works). I just tested that R's FFT has also a big speed difference for a long input with or without padding, so the padding should stay. Without padding R's fft() can be x100 slower for input lengths that are reasonable for Stan users. However it seems we should add mask correction https://etudes.tibonihoo.net/literate_musing/autocorrelations.html to remove the possible error from zero masking.
> 
> There is no hurry to add this as this kind of clearly non-stationary sequences are very unlikely and even then, e.g., ESS estimates are close enough to 1 so the error will not make anyone to think their chains converged when they didn't. So the fix would be mostly for purity and to people spending time wondering why the output doesn't match the true value (even if strange input is used).
> 
> Here's the code I think is correct
> 
> ```
> autocovariance <- function(x) {
>   N <- length(x)
>   M <- fft_next_good_size(N)
>   Mt2 <- 2 * M
>   yc <- x - mean(x)
>   yc <- c(yc, rep.int(0, Mt2 - N))
>   transform <- fft(yc)
>   ac <- fft(abs(transform)^2, inverse = TRUE)
>   # mask correction
>   (mask <- c(rep.int(1, N), rep.int(0, Mt2 - N)))
>   mask_correction <- fft(abs(fft(mask))^2, inverse = TRUE)
>   # use "biased" estimate as recommended by Geyer (1992)
>   # without mask correction
>   # ac <- Re(ac)[1:N] / (N^2 * 2)
>   # with mask correction
>   ac <- Re(ac)[1:N] / Re(mask_correction[1])
>   ac
> }
> ```
> 
> This can be tested against `acf(,type="covariance", lag.max=N)`, which seems to use FFT without zero padding (based on the speed behavior)

Hi all, thanks for the great discussion. I wanted to clarify something about statsmodels's implementation of the acov function.

Statsmodels's acov function is [here](https://github.com/statsmodels/statsmodels/blob/1a579185724949cf671705ae4f743b449977cd5f/statsmodels/tsa/stattools.py#L383-L527) which does use a [padding function](https://github.com/statsmodels/statsmodels/blob/1a579185724949cf671705ae4f743b449977cd5f/statsmodels/compat/scipy.py#L11-L53) as you described. So I'm not readily convinced that this is where the difference comes from.

**Comment by avehtari** [comment]
> Statsmodels's acov function is here which does use a padding function as you described. So I'm not readily convinced that this is where the difference comes from. Perhaps I might be too late in the overall development, though. Thanks again for all y'all's hard work!

I guess I was confused with many if-else's as when I read that code I tracked that the branches producing the result did not do masking. Anyway, it doesn't matter as the code in convergence.R is also not matching the Stan C++ code, and I think it's quite clear that convergence.R has a bug that was caused by trying to make a shortcut, but it doesn't have a big effect with the usual chains. The code I posted works fine.

**Comment by paul-buerkner** [comment]
Is the code below the correct one @avehtari? If so, I will replace the current autocovariance version by the below code.

> 
> 
> We can avoid the need to compute the mask effect (as it requires additional FFT and IFFT computations)
> This works as it's enough that we scale the autocovariances so that the first element matches var(x)*(N-1)/N
> 
> ```
> autocovariance <- function(x) {
>   N <- length(x)
>   # zero padding makes fft() much faster when N>1000
>   M <- fft_next_good_size(N)
>   Mt2 <- 2 * M
>   yc <- x - mean(x)
>   yc <- c(yc, rep.int(0, Mt2 - N))
>   # FFT based unnormalized autocovariances
>   ac <- Re(fft(abs(fft(yc))^2, inverse = TRUE)[1:N])
>   # use "biased" estimate as recommended by Geyer (1992)
>   # direct scaling with var(x) avoids need to compute "mask effect"
>   ac <- ac / ac[1] * var(x)*(N-1)/N
> }
> ```

**Comment by avehtari** [comment]
> Is the code below the correct one @avehtari? If so, I will replace the current autocovariance version by the below code.

Yes, although I recommend adding couple tests that would compare, e.g., to `acf(,type="covariance", lag.max=N)`

**Comment by paul-buerkner** [comment]
Done. Closing this issue.
